internal/util/version.go

```
@@ -1,57 +1,92 @@
 package util
 
 import (
+	"fmt"
 	"path/filepath"
-	"strings"
 
 	"github.com/edulinq/autograder/internal/log"
 )
 
 const (
 	UNKNOWN_VERSION  string = "???"
 	UNKNOWN_HASH     string = "????????"
-	VERSION_FILENAME string = "VERSION.txt"
+	VERSION_FILENAME string = "VERSION.json"
 	DIRTY_SUFFIX     string = "dirty"
+	UNKNOWN_API      int    = 0
 	HASH_LENGTH      int    = 8
 )
 
-func GetAutograderVersion() string {
+type Version struct {
+	Short   string `json:"short-version"`
+	Hash    string `json:"git-hash"`
+	IsDirty bool   `json:"is-dirty"`
+	Api     int    `json:"api-version"`
+}
+
+func (this Version) String() string {
+	fullVersion := this.Short + "-" + this.Hash
+	if this.IsDirty {
+		fullVersion = fullVersion + "-" + DIRTY_SUFFIX
+	}
+
+	return fullVersion
+}
+
+func MustGetAPIVersion() int {
+	version, err := versionFromJSON()
+	if err != nil {
+		log.Fatal("Failed to read the version from JSON file.", err)
+	}
+
+	return version.Api
+}
+
+func versionFromJSON() (Version, error) {
+	var version Version
+
 	versionPath := ShouldAbs(filepath.Join(ShouldGetThisDir(), "..", "..", VERSION_FILENAME))
 	if !IsFile(versionPath) {
-		log.Error("Version file does not exist.", log.NewAttr("path", versionPath))
-		return UNKNOWN_VERSION
+		return version, fmt.Errorf("Version path '%s' does not exist.", versionPath)
 	}
 
-	version, err := ReadFile(versionPath)
+	err := JSONFromFile(versionPath, &version)
 	if err != nil {
-		log.Error("Failed to read the version file.", err, log.NewAttr("path", versionPath))
-		return UNKNOWN_VERSION
+		return version, err
 	}
 
-	return strings.TrimSpace(version)
+	return version, nil
 }
 
-func GetAutograderFullVersion() string {
+func getGitInfo() (string, bool) {
```

You should return an error here (instead of logging) as well.